odoo/models.py

```diff
@@ -2072,16 +2069,87 @@ def _read_group_fill_temporal(self, data, groupby, aggregated_fields, annotated_
+        #     - fill_temporal_from: Apr (current month)
+        #     - fill_temporal_to:   Jul
+        # The result would be [Apr, May, Jun, Jul]
+        #
+        # 2. We want the same Forecast [Apr - Jul], but there is a chance that some
+        # records were forgotten in Jan, and we want to be able to reschedule
+        # them, but we don't need the fill_temporal to take effect from Jan.
+        #     - domain:             [(<= Jul)] (we removed the first bound)
+        #     - fill_temporal_from: Apr (current month)
+        #     - fill_temporal_to:   Jul
+        # The result would be [Jan, |Apr, May, Jun, Jul|]. We have the forecast
+        # with the fill_temporal, and every month preceding the forecast as long
+        # as there is at least one record in it (no empty month before the
+        # forecast).
+        tz = next((date.tzinfo for date in existing if date.tzinfo), False)
+        fill_from = self.env.context.get('fill_temporal_from')
+        fill_to = self.env.context.get('fill_temporal_to')
+        if fill_from:
+            fill_from = date_utils.start_of(datetime.datetime.strptime(fill_from, "%Y-%m-%d"), granularity)
+            if tz:
+                fill_from = tz.localize(fill_from)
+        elif existing_from:
+            fill_from = existing_from
+        else:
             return data
-
-        # assumption: existing data is sorted by field 'groupby_name'
-        first, last = existing[0], existing[-1]
+        if fill_to:
+            fill_to = date_utils.end_of(datetime.datetime.strptime(fill_to, "%Y-%m-%d"), granularity)
+            if tz:
+                fill_to = tz.localize(fill_to)
+        elif existing_to:
+            fill_to = existing_to
+        else:
+            fill_to = fill_from
+
+        if fill_to and fill_to < fill_from:
+            raise ValueError("Invalid context keys : from %s to %s" % (str(fill_from), str(fill_to)))
+
+        required_dates = list(date_utils.date_range(fill_from, fill_to, interval))
+
+        # the context key fill_temporal_min will be used as a way to guarantee a certain
+        # amount of returned groups for a read_group grouped on a {date, datetime} field.
+        # This will only work if there is at least one group to be used as reference (starting point)
+        # to create supplementary groups if necessary.
+        # example :
+        #   fill_temporal_min:  4
+        #   existing:           [2021-03-01, 2021-04-01]
+        #   interval:           1 month
+        # existing will be updated to contain 4 dates :
+        #   [2021-03-01, 2021-04-01, 2021-05-01, 2021-06-01]
+        min_groups = self.env.context.get('fill_temporal_min')
+        if min_groups and len(required_dates) < min_groups:
+            addon_from = interval + fill_to
+            addon_to = (min_groups - len(required_dates)) * interval + fill_to
+            required_dates += list(date_utils.date_range(addon_from, addon_to, interval))
```

Comment:
Can't this be computed beforehand, by checking the distance between `fill_from` and `fill_to` and adjusting `fill_to` accordingly? Or something along the lines of
```python
fill_to = max(fill_to, fill_from + min_groups * interval)
```

Reply:
This way we also don't need to reify `required_dates` to a list, we can just shove the result of the `date_range` into the set union.